pkg/tsdb/elasticsearch/field_namer.go

```diff
@@ -0,0 +1,170 @@
+package elasticsearch
+
+import (
+	"sort"
+	"strings"
+
+	"github.com/grafana/grafana-plugin-sdk-go/backend"
+	"github.com/grafana/grafana-plugin-sdk-go/data"
+)
+
+// nameFields applies naming logic to data frame fields based on query configuration
+func nameFields(queryResult backend.DataResponse, target *Query, keepLabelsInResponse bool) {
+	set := make(map[string]struct{})
+	frames := queryResult.Frames
+	for _, v := range frames {
+		for _, vv := range v.Fields {
+			if metricType, exists := vv.Labels["metric"]; exists {
+				if _, ok := set[metricType]; !ok {
+					set[metricType] = struct{}{}
+				}
+			}
+		}
+	}
+	metricTypeCount := len(set)
+	for _, frame := range frames {
+		if frame.Meta != nil && frame.Meta.Type == data.FrameTypeTimeSeriesMulti {
+			// if it is a time-series-multi, it means it has two columns, one is "time",
+			// another is "number"
+			valueField := frame.Fields[1]
+			fieldName := getFieldName(*valueField, target, metricTypeCount)
+			// If we  need to keep the labels in the response, to prevent duplication in names and to keep
+			// backward compatibility with alerting and expressions we use DisplayNameFromDS
+			if keepLabelsInResponse {
+				if valueField.Config == nil {
+					valueField.Config = &data.FieldConfig{}
+				}
+				valueField.Config.DisplayNameFromDS = fieldName
+				// If we don't need to keep labels (how frontend mode worked), we use frame.Name and remove labels
+			} else {
+				valueField.Labels = nil
+				frame.Name = fieldName
+			}
+		}
+	}
+}
+
+// getFieldName generates a field name based on the data field and target configuration
+func getFieldName(dataField data.Field, target *Query, metricTypeCount int) string {
+	metricType := dataField.Labels["metric"]
+	metricName := getMetricName(metricType)
+	delete(dataField.Labels, "metric")
```

Comment:
`dataField.Labels` is a `map`, and therefore a mutable reference. I'm agnostic towards making this change right now, but we might want to make a copy and work on that instead of mutating the reference below.